<a href="https://colab.research.google.com/github/pelinbalci/TF_Intro/blob/main/lecture_3_convolutions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import tensorflow as tf
import numpy as np



```
model = tf.keras.models.Sequential([
tf.keras.layers.Conv2D(),
tf.keras.layers.MaxPooling2D(),
tf.keras.layers.Conv2D(),
tf.keras.layers.MaxPooling2D(),
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(),
tf.keras.layers.Dense()
])
```

lets write the parameters:

```
tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28,28,1),
```

with model summary you can see the changes:

64 is the number of filters for this layer. 

3*3 is the size of the filter.

input_shape is the shape of the input we fed in which is 28*28 with 1 byte color depth


```
tf.keras.layers.MaxPooling2D(2,2)
```

2*2 defines the chunks to pool. 





In [3]:
model = tf.keras.models.Sequential([
tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28,28,1)),
tf.keras.layers.MaxPooling2D(2,2),
tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
tf.keras.layers.MaxPooling2D(2,2),
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(128, activation='relu'),
tf.keras.layers.Dense(10, activation='softmax')
])

In [4]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 26, 26, 64)        640       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 64)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 11, 11, 64)        36928     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
flatten (Flatten)            (None, 1600)              0         
_________________________________________________________________
dense (Dense)                (None, 128)               204928    
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1

(None, 26, 26, 64), since we don't use padding here. 

Each filter will learn 9 values for the filter coefficients + bias --> 64 filters have 640 learnable parameters.


pooling reduce the dimensionality by half. but there is not any learnable params here. 


3*3 filter then reduces 13*13 to 11*11 by removing a pixel border like before. 

max pool half of it --> 5*5

so we flatten: 64*5*5 = 1600

dense network = 1600* 128 = 204800


final layer: 128*10 + 10 bias = 1290 params


In [6]:
import tensorflow as tf
import numpy as np

In [7]:
#load data
mnist = tf.keras.datasets.fashion_mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train/255, x_test/255

4423680/4422102 [==============================] - 0s 0us/step


In [10]:
# create model 
#note: without giving input shape model summary gives error
model = tf.keras.models.Sequential([
tf.keras.layers.Flatten(input_shape=(28,28)),
tf.keras.layers.Dense(512, activation='relu'),
tf.keras.layers.Dense(10, activation='softmax')                             
]
)

model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 512)               401920    
_________________________________________________________________
dense_7 (Dense)              (None, 10)                5130      
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


In [11]:
#define optimizer and train the model and get test loss
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(x_train, y_train, epochs=5)
test_loss = model.evaluate(x_test, y_test)


Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 15.1962 - accuracy: 0.7335
Epoch 2/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.5378 - accuracy: 0.8154
Epoch 3/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.4956 - accuracy: 0.8311
Epoch 4/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.4729 - accuracy: 0.8369
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.5262 - accuracy: 0.8335


Before using conv layers we need to reshape the input: 

the input shape should be: observationsize, 28*28 pixels, 1 color depth --> 60000, 28, 28, 1

In [27]:
# load the input
mnist = tf.keras.datasets.fashion_mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# x_train.shape  # (60000, 28, 28)

# reshape data
x_train = x_train.reshape(x_train.shape[0], x_train.shape[1], x_train.shape[2], 1)
x_test = x_test.reshape(x_test.shape[0], x_test.shape[1], x_test.shape[2], 1)

# normalize data 
min_val, max_val = x_train[1].min(), x_train[1].max()
x_train, x_test = x_train / max_val, x_test/max_val

# create model
model = tf.keras.models.Sequential([
tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28,28,1)),
tf.keras.layers.MaxPool2D(2,2),
tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
tf.keras.layers.MaxPool2D(2,2),
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(128, activation='relu'),
tf.keras.layers.Dense(10, activation='softmax')                                   
])
model.summary()

# define optimizer
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics='accuracy')

# train the model
model.fit(x_train, y_train, epochs=5)

# get test results 
model.evaluate(x_test, y_test)

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_9 (Conv2D)            (None, 26, 26, 64)        640       
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 13, 13, 64)        0         
_________________________________________________________________
conv2d_10 (Conv2D)           (None, 11, 11, 64)        36928     
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 5, 5, 64)          0         
_________________________________________________________________
flatten_7 (Flatten)          (None, 1600)              0         
_________________________________________________________________
dense_14 (Dense)             (None, 128)               204928    
_________________________________________________________________
dense_15 (Dense)             (None, 10)               

[0.2625056803226471, 0.9067000150680542]